tools/insn_freq_explorer.c

```diff
@@ -0,0 +1,265 @@
+}
+
+static int cmp_asc(const void *a, const void *b)
+{
+    return ((rv_insn_hist_t *) a)->freq - ((rv_insn_hist_t *) b)->freq;
+}
+
+/* used to adjust the length of histogram bar */
+static unsigned short get_win_max_col()
+{
+    struct winsize terminal_size;
+    ioctl(STDOUT_FILENO, TIOCGWINSZ, &terminal_size);
+
+    return terminal_size.ws_col;
+}
+
+/*      No.     instruction     percentages [frequency] histogram */
+static const char *fmt = "%3d. %-10s%5.2f%% [%-10zu] %s\n";
+
+/* get columns used by the fmt string */
+static unsigned short get_used_col()
+{
+    unsigned short used_col = 0;
+
+    used_col += 3;  /* width = 3 */
+    used_col += 1;  /* . */
+    used_col += 1;  /* single space */
+    used_col += 10; /* width = 10 */
+    used_col += 5;  /* width = 5 */
+    used_col += 1;  /* % */
+    used_col += 1;  /* single space */
+    used_col += 1;  /* [ */
+    used_col += 10; /* width = 10 */
+    used_col += 1;  /* ] */
+    used_col += 1;  /* single space */
+
+    return used_col;
+}
+
+static char *gen_hist_bar(char *hist_bar,
+                          size_t insn_freq,
+                          size_t max_insn_freq,
+                          unsigned short max_col,
+                          unsigned short used_col)
+{
+    int hist_bar_len = insn_freq * (max_col - used_col) / max_insn_freq;
+    for (int i = 0; i < hist_bar_len; i++) {
+        hist_bar[i] = '*';
+    }
+    hist_bar[hist_bar_len] = 0;
+    return hist_bar;
+}
+
+static void print_usage(const char *filename)
+{
+    fprintf(stderr,
+            "RV32I[MACF] Instruction Frequency Explorer which loads an ELF "
+            "file to execute.\n"
+            "Usage: %s [option] [elf_file_path]\n"
+            "available option: -a, print the histogram in ascending order\n",
+            filename);
+}
+
+static bool parse_args(int argc, char **args)
+{
+    int opt;
+
+    while ((opt = getopt(argc, args, optstr)) != -1) {
+        switch (opt) {
+        case 'a':
+            ascending_order = true;
+            break;
+        default:
+            return false;
+        }
+    }
+
+    elf_prog = args[optind];
+
+    return true;
+}
+
+
+int main(int argc, char *args[])
+{
+    if (!parse_args(argc, args) || !elf_prog) {
+        print_usage(args[0]);
+        return 1;
+    }
+
+    elf_t *e = elf_new();
+    if (!elf_open(e, elf_prog)) {
+        fprintf(stderr, "elf_open failed\n");
+        return 1;
+    }
+
+    struct Elf32_Ehdr *hdr = (struct Elf32_Ehdr *) e->hdr;
+    struct Elf32_Shdr **shdrs = get_elf_section_headers(e);
+    if (!shdrs) {
+        fprintf(stderr, "malloc for section headers failed\n");
+        return 1;
+    }
+    uint8_t *elf_first_byte = (uint8_t *) e->raw_data;
+    size_t total_insn = 0;
+    rv_insn_t ir;
+    bool res;
+
+    for (int i = 0; i < hdr->e_shnum; i++) {
+        struct Elf32_Shdr *shdr = (struct Elf32_Shdr *) shdrs + i;
+
+        if (!strcmp(get_sh_type(shdr->sh_type), "PROGBITS") &&
```

Comment:
Avoid comparing strings. Instead, check the values.

Reply:
Rewrite as following:
```c
    if ( ... COND_TO_SKIP_THIS_ITERATION ...)
        continue;
     uint32_t *exec_start_addr =
                (uint32_t *) ((uint8_t *) elf_first_byte + shdr->sh_offset);
    uint32_t *exec_end_addr =
    (uint32_t *) ((uint8_t *) exec_start_addr + shdr->sh_size);
    ...
```

Reply:
See https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md (Avoid redundant else)